File: python2/closure.py
def closure2(function,  iterable): return closureI(close2(function), iterable)
def closure2S(function, iterable): return closureS(close2(function), iterable)

def close2(function):
    def _close2(iter1, iter2):
	for a in iter1:
	    for b in iter2:
		yield function(a,b)
		yield function(b,a)
    return _close2

def closure2A(function, iterable):
    # For use when `function` is associative
    start = frozenset(iterable)
    def _close2A(old, new): return (function(y,x) for x in new for y in start)
    return closureI(_close2A, start)

def closure2AGens(function, generators):
    generators = frozenset(generators)
    #elems = {x: [x] for x in generators}
    elems = {}
    for x in generators: elems[x] = [x]
    newQueue = generators
    while newQueue:
	(oldQueue, newQueue) = (newQueue, [])
	for x in oldQueue:
	    for y in generators:
		z = function(x,y)
		if z not in elems:
		    elems[z] = elems[x] + [y]
		    newQueue.append(z)
    return elems

def closure1m(function, iterable):
    # `function` must have type `x -> iterable`.
    def _close1m(old, new): return (y for x in new for y in function(x))
    return closureI(_close1m, iterable)

def closure1mS(function, iterable):
    # `function` must have type `x -> iterable`.
    def _close1m(old, new): return (y for x in new for y in function(x))
    return closureS(_close1m, iterable)

def closureI(function, iterable):
    # `function` must have type `(iterable /* old */, iterable /* new */) ->
    # iterable`.
    seen = set(iterable)
    new = list(seen)
    for x in new:
	yield x
    while new:
	new = list(view(function(seen.copy(), new), seen))
	for x in new:
	    yield x

def closureS(function, iterable):
    # `function` must have type `(iterable /* old */, iterable /* new */) ->
    # iterable`.
    seen = set(iterable)
    new = list(seen)
    while new:
	new = list(view(function(seen.copy(), new), seen))
    return seen

def view(iterable, seen):
    for x in iterable:
	if x not in seen:
	    seen.add(x)
	    yield x

#def view(iterable, seen):
#    seen_add = seen.add  # Apparently this is a worthwhile optimization.
#    return (x for x in iterable if x not in seen and not seen_add(x))

File: haskell/Closure.hs
module Closure (
  -- * Closure functions
  closure2, closure2', closure2'',
  closure2A, closure2A',
  closure1m, closure1m',
  -- * Utility functions
  closureL, closureS, close2, view
 ) where
 import Data.Set (Set, fromList, toList, member, insert)

 closure2 :: Ord a => (a -> a -> a) -> [a] -> [a]
 closure2 f start = closureL (close2 f) (start, fromList start)

 closure2' :: Ord a => (a -> a -> a) -> [a] -> Set a
 closure2' f start = closureS (close2 f) (start, fromList start)

 closure2'' :: Ord a => (a -> a -> a) -> Set a -> Set a
 closure2'' f start = closureS (close2 f) (toList start, start)

 closure2A :: Ord a => (a -> a -> a) -> [a] -> [a]
 -- For use when @f@ is associative
 closure2A f start = closure1m (\x -> map (`f` x) start) start

 closure2A' :: Ord a => (a -> a -> a) -> [a] -> Set a
 -- For use when @f@ is associative
 closure2A' f start = closure1m' (\x -> map (`f` x) start) start

 closure1m :: Ord a => (a -> [a]) -> [a] -> [a]
 closure1m f start = closureL (\xs _ -> concatMap f xs) (start, fromList start)

 closure1m' :: Ord a => (a -> [a]) -> [a] -> Set a
 closure1m' f start = closureS (\xs _ -> concatMap f xs) (start, fromList start)

 closureL :: Ord a => ([a] -> Set a -> [a]) -> ([a], Set a) -> [a]
 closureL _ ([],  _) = []
 closureL f (new, seen) = new ++ closureL f (view (f new seen) seen)

 closureS :: Ord a => ([a] -> Set a -> [a]) -> ([a], Set a) -> Set a
 closureS _ ([],  seen) = seen
 closureS f (new, seen) = closureS f (view (f new seen) seen)

 close2 :: Ord a => (a -> a -> a) -> [a] -> Set a -> [a]
 -- Reconsider this function's name; "bisetmap"? "doublemap"? "ambimap"?
 close2 f new seen = concat [[f a b, f b a] | a <- toList seen, b <- new]

 -- |Given a list @xs@ and a set @s@ of "seen" values, @view xs s@ returns a
 -- list of (non-duplicate) values in @xs@ but not in @s@ and an updated
 -- version of @s@ that includes the values in @xs@.
 view :: Ord a => [a] -> Set a -> ([a], Set a)
 view [] seen = ([], seen)
 view (x:xs) seen | member x seen = view xs seen
 view (x:xs) seen = x &: view xs (insert x seen)
  where z &: (zs, y) = (z:zs, y)
